parsers/VN.py

```diff
@@ -0,0 +1,255 @@
+
+
+def fetch_consumption(
+    zone_key: str,
+    session: Session = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> Union[Dict[str, Any], List[Dict[str, Any]]]:
+
+    request_latest = target_datetime is None
+
+    if request_latest:
+        fetched_day, fetched_data = fetch_latest_data("consumption", session)
+
+    else:
+        # TODO API bug for past data:
+        # data for given date can only be retrieved until the current hour.
+        fetched_day, fetched_data = fetch_target_data(
+            "consumption", target_datetime, session
+        )
+
+    # [0] = load of Northern Vietnam
+    # [1] = load of Central Vietnam
+    # [2] = load of Southern Vietnam
+    # [3] = load of Vietnam (total)
+    per_zone_load = {
+        "VN-N": fetched_data["data"][0],
+        "VN-C": fetched_data["data"][1],
+        "VN-S": fetched_data["data"][2],
+        "VN": fetched_data["data"][3],
+    }
+
+    result_list = []
+
+    for i in range(len(per_zone_load[zone_key])):
+        # Convert to datetime. Data is for [0:30, 1:00, 1:30, ..., 23:30, 24:00]
+        # Or only up to latest available time
+        i_datetime = fetched_day.replace(
+            hour=int((i + 1) / 2), minute=(0 if i % 2 else 30)
+        ).floor("minute")
+
+        # Data also goes ~2h into the future! Seems to be some kind of prediction, since these
+        # future values are changing until their valid time is reached.
+        if request_latest and i_datetime > arrow.now(tz=tz):
+            continue
+
+        result_list.append(
+            {
+                "datetime": i_datetime.datetime,
+                "consumption": per_zone_load[zone_key][i],
+                "zoneKey": zone_key,
+                "source": "nldc.evn.vn",
+            }
+        )
+
+    if not len(result_list):
+        raise ParserException(
+            parser="VN.py",
+            message=f"No valid consumption data for requested day found.",
+        )
+
+    return result_list
+
+
+def filter_prices(per_zone_price):
+    """
+    Filter the prices for each zone is_price_valid lambda function. If they seem wrong, replace them with None.
+    """
+    for z, prices in per_zone_price.items():
+        per_zone_price[z] = [(p if is_price_valid(p) else None) for p in prices]
+
+    return per_zone_price
+
+
+def compute_national_average(per_zone_price):
+    """
+    Prices only given for each of the 3 zones, compute mean of the three zones as national average.
+    Also handle if some of these are None.
+    """
+
+    averaged_vn_prices = []
+    # average of all non-nan items
+    for i in range(len(per_zone_price["VN-N"])):
+        zone_prices = [
+            ps[i] for zone, ps in per_zone_price.items() if ps[i] is not None
+        ]
+
+        if len(zone_prices):
+            averaged_vn_prices.append(mean(zone_prices))
+        else:
+            averaged_vn_prices.append(None)
+
+    return averaged_vn_prices
```

Comment:
Hmm, not sure I like that we are creating data here.

I think it would be best to leave it out and just add price data to the subzones.

Reply:
Done. 09b29ec